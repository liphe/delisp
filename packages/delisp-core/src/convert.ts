import { printHighlightedExpr } from "./error-report";
import {
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";
import {
  Declaration,
  Expression,
  LambdaList,
  SIdentifier,
  SLetBinding,
  Syntax
} from "./syntax";
import { last } from "./utils";

import {
  checkUserDefinedTypeName,
  convert as convertType
} from "./convert-type";
import { parseRecord } from "./convert-utils";
import { listTypeVariables } from "./type-utils";
import { TypeWithWildcards } from "./type-wildcards";

const conversions: Map<string, (expr: ASExprList) => Expression> = new Map();
const toplevelConversions: Map<
  string,
  (expr: ASExprList) => Declaration
> = new Map();

function defineConversion(name: string, fn: (expr: ASExprList) => Expression) {
  conversions.set(name, fn);
}

function defineToplevel(name: string, fn: (expr: ASExprList) => Declaration) {
  toplevelConversions.set(name, fn);
}

function parseBody(anchor: ASExpr, exprs: ASExpr[]): Expression[] {
  if (exprs.length === 0) {
    throw new Error(
      printHighlightedExpr(`body can't be empty`, anchor.location, true)
    );
  }
  return exprs.map(convertExpr);
}

//
// The format of lambda lists are (a b c ... &rest z)
//

function parseLambdaList(ll: ASExpr): LambdaList {
  if (ll.tag !== "list") {
    throw new Error(
      printHighlightedExpr("Expected a list of arguments", ll.location)
    );
  }

  ll.elements.forEach(arg => {
    if (arg.tag !== "symbol") {
      throw new Error(
        printHighlightedExpr(
          "A list of arguments should be made of symbols",
          arg.location
        )
      );
    }
  });

  const symbols = ll.elements as ASExprSymbol[];

  // Check for duplicated arguments
  symbols.forEach((arg, i) => {
    const duplicated = symbols.slice(i + 1).find(a => a.name === arg.name);
    if (duplicated) {
      throw new Error(
        printHighlightedExpr(
          "There is another argument with the same name",
          duplicated.location
        )
      );
    }
  });

  return {
    positionalArgs: symbols.map(convertSymbol),
    location: ll.location
  };
}

defineConversion("lambda", expr => {
  const [lambda, ...args] = expr.elements;
  const lastExpr = last([lambda, ...args]) as ASExpr; // we kj

  if (args.length === 0) {
    throw new Error(
      printHighlightedExpr(
        `'lambda' is missing the argument list`,
        lastExpr.location,
        true
      )
    );
  }

  const body = parseBody(lastExpr, args.slice(1));

  return {
    tag: "function",
    lambdaList: parseLambdaList(args[0]),
    body,
    location: expr.location,
    info: {}
  };
});

defineConversion("if", expr => {
  if (expr.elements.length !== 4) {
    const lastExpr = last(expr.elements) as ASExpr; // we know it is not empty!
    throw new Error(
      printHighlightedExpr(
        `'if' needs exactly 3 arguments, got ${expr.elements.length}`,
        lastExpr.location,
        true
      )
    );
  }
  const [, conditionForm, consequentForm, alternativeForm] = expr.elements;
  return {
    tag: "conditional",
    condition: convertExpr(conditionForm),
    consequent: convertExpr(consequentForm),
    alternative: convertExpr(alternativeForm),
    location: expr.location,
    info: {}
  };
});

function parseLetBindings(bindings: ASExpr): SLetBinding[] {
  if (bindings.tag !== "map") {
    throw new Error(
      printHighlightedExpr(`'let' bindings should be a map`, bindings.location)
    );
  }

  return bindings.fields.map(field => ({
    variable: convertSymbol(field.label),
    value: convertExpr(field.value)
  }));
}

defineConversion("let", expr => {
  const [_let, ...args] = expr.elements;
  const lastExpr = last([_let, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new Error(
      printHighlightedExpr(
        `'let' is missing the bindings`,
        lastExpr.location,
        true
      )
    );
  }

  const [rawBindings, ...rawBody] = args;

  return {
    tag: "let-bindings",
    bindings: parseLetBindings(rawBindings),
    body: parseBody(lastExpr, rawBody),
    location: expr.location,
    info: {}
  };
});

defineConversion("the", expr => {
  const [_the, ...args] = expr.elements;
  const lastExpr = last([_the, ...args]) as ASExpr; // we know it is not empty!

  if (args.length === 0) {
    throw new Error(
      printHighlightedExpr(
        `'the' is missing the type and value`,
        lastExpr.location,
        true
      )
    );
  }
  if (args.length === 1) {
    throw new Error(
      printHighlightedExpr(
        `'the' is missing the expression`,
        lastExpr.location,
        true
      )
    );
  }

  if (args.length > 2) {
    throw new Error(
      printHighlightedExpr(
        `Too many arguments. 'the' should take two arguments.`,
        args[2].location,
        true
      )
    );
  }

  const [t, value] = args;

  return {
    tag: "type-annotation",
    typeWithWildcards: new TypeWithWildcards(convertType(t)),
    value: convertExpr(value),
    location: expr.location,
    info: {}
  };
});

defineToplevel("define", expr => {
  const [define, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([define, ...args]) as ASExpr;
    throw new Error(
      printHighlightedExpr(
        `'define' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable, value] = args;

  if (variable.tag !== "symbol") {
    throw new Error(
      printHighlightedExpr("'define' expected a symbol", variable.location)
    );
  }

  return {
    tag: "definition",
    variable: convertSymbol(variable),
    value: convertExpr(value),
    location: expr.location
  };
});

defineToplevel("export", expr => {
  const [exp, ...args] = expr.elements;

  if (args.length !== 1) {
    const lastExpr = last([exp, ...args]) as ASExpr;
    throw new Error(
      printHighlightedExpr(
        `'export' needs exactly 1 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [variable] = args;

  if (variable.tag !== "symbol") {
    throw new Error(
      printHighlightedExpr("'export' expected a symbol", variable.location)
    );
  }

  return {
    tag: "export",
    value: convertSymbol(variable),
    location: expr.location
  };
});

defineToplevel("type", expr => {
  const [typeOp, ...args] = expr.elements;

  if (args.length !== 2) {
    const lastExpr = last([typeOp, ...args]) as ASExpr;
    throw new Error(
      printHighlightedExpr(
        `'type' needs exactly 2 arguments, got ${args.length}`,
        lastExpr.location,
        true
      )
    );
  }

  const [name, definition] = args;

  if (name.tag !== "symbol") {
    throw new Error(
      printHighlightedExpr("'type' expected a symbol as a name", name.location)
    );
  }

  checkUserDefinedTypeName(name);

  const definitionType = convertType(definition);
  if (listTypeVariables(definitionType).length > 0) {
    throw new Error(
      printHighlightedExpr("Type variable out of scope", definition.location)
    );
  }

  return {
    tag: "type-alias",
    alias: convertSymbol(name),
    definition: definitionType,
    location: expr.location
  };
});

function convertList(list: ASExprList): Expression {
  if (list.elements.length === 0) {
    throw new Error(
      printHighlightedExpr("Empty list is not a function call", list.location)
    );
  }

  const [first] = list.elements;

  const convertSpecialForm =
    first.tag === "symbol" ? conversions.get(first.name) : undefined;

  if (convertSpecialForm) {
    return convertSpecialForm(list);
  } else {
    const [fn, ...args] = list.elements;
    return {
      tag: "function-call",
      fn: convertExpr(fn),
      args: args.map(convertExpr),
      location: list.location,
      info: {}
    };
  }
}

function convertVector(list: ASExprVector): Expression {
  return {
    tag: "vector",
    values: list.elements.map(a => convertExpr(a)),
    location: list.location,
    info: {}
  };
}

function convertMap(map: ASExprMap): Expression {
  const { fields, tail } = parseRecord(map);

  return {
    tag: "record",
    fields: fields.map(f => ({
      label: convertSymbol(f.label),
      value: convertExpr(f.value)
    })),
    extends: tail && convertExpr(tail),
    location: map.location,
    info: {}
  };
}

function convertSymbol(expr: ASExprSymbol): SIdentifier {
  return {
    tag: "identifier",
    name: expr.name,
    location: expr.location,
    info: {}
  };
}

export function convertExpr(expr: ASExpr): Expression {
  switch (expr.tag) {
    case "number":
      return { ...expr, info: {} };
    case "string":
      return { ...expr, info: {} };
    case "symbol":
      return convertSymbol(expr);
    case "list":
      return convertList(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
  }
}

export function convert(expr: ASExpr): Syntax {
  if (expr.tag === "list") {
    if (expr.elements.length === 0) {
      throw new Error(
        printHighlightedExpr("Empty list is not a function call", expr.location)
      );
    }

    const [first] = expr.elements;
    const convertDeclaration =
      first.tag === "symbol" ? toplevelConversions.get(first.name) : undefined;

    if (convertDeclaration) {
      return convertDeclaration(expr);
    }
  }

  return convertExpr(expr);
}
