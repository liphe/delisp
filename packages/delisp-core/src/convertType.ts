import { printHighlightedExpr } from "./error-report";
import { ASExpr, ASExprList, ASExprSymbol } from "./sexpr";
import { Type } from "./types";

function convertSymbol(expr: ASExprSymbol): Type {
  switch (expr.name) {
    case "number":
      return { type: "number" };
    case "string":
      return { type: "string" };
    default:
      return { type: "type-variable", name: expr.name };
  }
}

function convertList(expr: ASExprList): Type {
  const [first, from, to] = expr.elements;

  if (first.type !== "symbol" || first.name !== "->") {
    throw new Error(printHighlightedExpr("Expected -> as first symbol", expr));
  }

  if (expr.elements.length !== 3) {
    throw new Error(printHighlightedExpr("Expected 3 elements", expr));
  }

  if (from.type !== "list") {
    throw new Error(
      printHighlightedExpr("Expected second element to be a list", expr)
    );
  }

  return {
    type: "function",
    from: from.elements.map(convert),
    to: convert(to)
  };
}

export function convert(expr: ASExpr): Type {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr));
  }
}
