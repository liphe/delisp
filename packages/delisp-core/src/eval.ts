//
// Evaluation
//

import * as vm from "vm";
import { compileToString, Environment, moduleEnvironment } from "./compiler";
import primitives from "./primitives";
import { Module, Syntax } from "./syntax";
import { mapObject } from "./utils";

import {
  caseTag,
  matchTag,
  pair,
  fst,
  snd,
  values,
  primaryValue,
  bindPrimaryValue
} from "@delisp/runtime";

export function createContext() {
  const sandbox = {
    env: mapObject(primitives, p => p.value),
    console,
    // Primitives
    caseTag,
    matchTag,
    pair,
    fst,
    snd,
    values,
    primaryValue,
    bindPrimaryValue
  };
  vm.createContext(sandbox);
  return sandbox;
}

export function evaluate(
  syntax: Syntax,
  env: Environment,
  context = createContext()
): unknown {
  const code = compileToString(syntax, env);
  const result = vm.runInContext(code, context);
  return result;
}

export function evaluateModule(m: Module, context = createContext()): void {
  const env = moduleEnvironment(m, { definitionContainer: "env" });
  m.body.forEach(s => {
    return evaluate(s, env, context);
  });
}
