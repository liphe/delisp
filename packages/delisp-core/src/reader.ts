//
// reader.ts --- The reader for Deisp
//

//
// S-Expression parser
//

import {
  alternatives,
  atLeastOne,
  character,
  delimited,
  delimitedMany,
  endOfInput,
  many,
  Parser,
  until
} from "./parser-combinators";

import { ASExpr } from "./sexpr";

const spaces = many(
  alternatives(character(" "), character("\t"), character("\n"))
);

function spaced<A>(x: Parser<A>) {
  return delimited(spaces, x, spaces);
}

const alphanumeric = character().chain(char => {
  if (/[a-zA-Z0-9]/.test(char)) {
    return Parser.of(char);
  } else {
    return Parser.fail("an alphanumeric character");
  }
});

const digit = character().chain(char => {
  if (/[0-9]/.test(char)) {
    return Parser.of(char);
  } else {
    return Parser.fail("a digit");
  }
});

//
// Number
//

const unsignedInteger: Parser<number> = atLeastOne(digit).map(digits =>
  parseInt(digits.join(""), 10)
);
const negativeInteger: Parser<number> = character("-")
  .chain(_ => unsignedInteger)
  .map(num => -num);

const numberP: Parser<ASExpr> = alternatives(negativeInteger, unsignedInteger)
  .map(
    (val, location): ASExpr => ({
      type: "number",
      value: val,
      location
    })
  )
  .description("number");

//
// String
//
const doubleQuote = character('"').description("double quote");

const stringChar = character()
  .chain(char => {
    if (char === '"') {
      return Parser.fail("Not inside string");
    }
    if (char === "\\") {
      return alternatives(
        character("n").map(_ => "\n"),
        character("\\")
      ).description("escaped string character");
    }
    return Parser.of(char);
  })
  .description("string character");

const stringP = delimitedMany(doubleQuote, stringChar, doubleQuote)
  .map(
    (chars, location): ASExpr => ({
      type: "string",
      value: chars.join(""),
      location
    })
  )
  .description("string");

//
// Symbol
//

const symbol: Parser<ASExpr> = atLeastOne(
  alternatives(
    alphanumeric,
    character("!"),
    character("@"),
    character("#"),
    character("$"),
    character("%"),
    character("^"),
    character("&"),
    character("*"),
    character("<"),
    character(">"),
    character("+"),
    character("-"),
    character("/"),
    character("~"),
    character("?")
  )
)
  .map(
    (chars, location): ASExpr => ({
      type: "symbol",
      name: chars.join(""),
      location
    })
  )
  .description("symbol");

//
// Lists & S-Expressions
//

const atom: Parser<ASExpr> = alternatives(numberP, stringP, symbol).description(
  "atom"
);

const leftParen = character("(").description("open parenthesis");
const rightParen = character(")").description("close parenthesis");

function list(x: Parser<ASExpr>): Parser<ASExpr> {
  return delimitedMany(leftParen, x, spaces.then(rightParen))
    .map(
      (elements, location): ASExpr => ({
        type: "list",
        elements,
        location
      })
    )
    .description("list");
}

const reportUnmatched: Parser<{}> = Parser.lookahead(rightParen).chain(
  closed => {
    if (closed) {
      return Parser.fail("Unmatched closed parenthesis");
    } else {
      return Parser.of({});
    }
  }
);

const sexpr: Parser<ASExpr> = spaced(
  reportUnmatched.then(atom.or(() => list(sexpr)))
);

const sexprs: Parser<ASExpr[]> = until(endOfInput, sexpr);

export function readAllFromString(str: string): ASExpr[] {
  return sexprs.parse(str);
}

//
// Parser a Delisp expression from a string
//
export function readFromString(str: string): ASExpr {
  return sexpr.parse(str);
}
