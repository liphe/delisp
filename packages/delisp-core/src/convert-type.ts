import { capitalize } from "./utils";
import { ConvertError, parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import { readFromString } from "./reader";
import { last } from "./utils";

import {
  isSymbolOfName,
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";

import {
  emptyRow,
  TConstant,
  TVar,
  Type,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tMultiValuedFunction,
  tValues,
  tCases,
  tEffect,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid,
  tNone,
  tcArrow,
  tcStar,
  TypeSchema
} from "./types";

import { TypeWithWildcards } from "./type-wildcards";

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return /^[A-Z]/.test(expr.name);
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new ConvertError(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): TVar | TConstant {
  switch (expr.name) {
    case "->":
      return tcArrow;
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "none":
      return tNone;
    case "*":
      return tcStar;
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertEffect(effects: ASExpr[]): Type {
  const labels = effects.map(e => {
    if (e.tag !== "symbol") {
      throw new ConvertError(
        printHighlightedExpr(`not a valid effect`, e.location)
      );
    }
    return e.name;
  });
  const last = labels.slice(-2);
  if (last.length === 2 && last[0] === "|") {
    return tEffect(labels.slice(0, -2), tVar(last[1]));
  } else {
    return tEffect(labels);
  }
}

function convertCases(_op: ASExpr, args: ASExpr[]): Type {
  const lastArg = last(args);
  let tail: ASExpr | undefined;

  if (lastArg && lastArg.tag === "symbol" && !lastArg.name.startsWith(":")) {
    tail = args.pop();
  }

  function parseCase(c: ASExpr): { label: string; type: Type } {
    if (c.tag === "symbol" && c.name.startsWith(":")) {
      return { label: c.name, type: tVoid };
    } else if (c.tag === "list") {
      if (c.elements.length !== 2) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected exaclty two elements`,
            c.location
          )
        );
      }
      const [keyword, innerType] = c.elements;
      if (keyword.tag !== "symbol" || !keyword.name.startsWith(":")) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected valid keyword`,
            keyword.location
          )
        );
      }
      return { label: keyword.name, type: convert_(innerType) };
    } else {
      throw new ConvertError(printHighlightedExpr(`Invalid case`, c.location));
    }
  }

  return tCases(args.map(parseCase), tail && convert_(tail));
}

function convertValues(expr: ASExpr): Type[] {
  // Check if X is the symbol VALUES
  if (
    expr.tag === "list" &&
    expr.elements.length > 0 &&
    isSymbolOfName(expr.elements[0], "values")
  ) {
    return expr.elements.slice(1).map(convert_);
  } else {
    return [convert_(expr)];
  }
}

function convertList(expr: ASExprList): Type {
  const [op, ...args] = expr.elements;
  if (op.tag === "symbol" && op.name === "effect") {
    return convertEffect(args);
  } else if (op.tag === "symbol" && op.name === "cases") {
    return convertCases(op, args);
  } else if (op.tag === "symbol" && op.name === "->") {
    if (args.length < 2) {
      throw new ConvertError(`Missing some arguments!`);
    }
    const fnargs = args.slice(0, -2);
    const fneffect = args[args.length - 2];
    const fnout = args[args.length - 1];
    return tMultiValuedFunction(
      fnargs.map(convert_),
      convert_(fneffect),
      tValues(convertValues(fnout))
    );
  } else {
    const opType = convert_(op);
    return tApp(opType, ...args.map(convert_));
  }
}

function convertVector(expr: ASExprVector): Type {
  if (expr.elements.length !== 1) {
    throw new ConvertError(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert_(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Type {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert_(value)
    })),
    tail ? convert_(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
function convert_(expr: ASExpr): Type {
  switch (expr.tag) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new ConvertError(
        printHighlightedExpr("Not a valid type", expr.location)
      );
  }
}

function convertToTypeWithWildcards(expr: ASExpr): TypeWithWildcards {
  return new TypeWithWildcards(convert_(expr));
}

export { convertToTypeWithWildcards as convert };

export function readType(source: string): TypeSchema {
  const t = convertToTypeWithWildcards(readFromString(source));
  return t.generalize();
}
