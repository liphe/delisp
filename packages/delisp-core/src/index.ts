import { convert as convertSyntax } from "./convert";

import { readFromString } from "./reader";

export { collectConvertErrors } from "./convert";

export {
  compileToString,
  compileModuleToString,
  moduleEnvironment
} from "./compiler";
export { createContext, evaluate, evaluateModule } from "./eval";
export {
  inferType,
  inferModule,
  inferExpressionInModule,
  defaultEnvironment
} from "./infer";
export { Type } from "./types";
export { printType } from "./type-printer";
export {
  isDeclaration,
  isDefinition,
  isExpression,
  isExport,
  isTypeAlias
} from "./syntax";
export { default as primitives } from "./primitives";

export { printHighlightedExpr } from "./error-report";

export function readSyntax(source: string) {
  return convertSyntax(readFromString(source));
}

export { findSyntaxByOffset, findSyntaxByRange } from "./syntax-utils";

export {
  createModule,
  readModule,
  addToModule,
  removeModuleDefinition,
  removeModuleTypeDefinition,
  moduleDefinitions,
  moduleDefinitionByName
} from "./module";

export { generateTSModuleDeclaration } from "./typescript-generation";

export { Encoder } from "./prettier";
export { pprintAs, pprint, pprintModuleAs, pprintModule } from "./printer";
