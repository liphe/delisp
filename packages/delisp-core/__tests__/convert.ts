import { convert, collectConvertErrors } from "../src/convert";
import { readFromString } from "../src/reader";

describe("Convert", () => {
  describe("Error messages", () => {
    function compileError(str: string): string {
      const sexpr = readFromString(str);
      const syntax = convert(sexpr);
      const errors = collectConvertErrors(syntax);
      if (errors.length === 0) {
        throw new Error(`FATAL: EXPRESSION ${str} DID NOT FAIL TO COMPILE.`);
      }
      return errors.join("\n\n");
    }

    it("generate a nice error for some basic invalid syntax", () => {
      expect(compileError("()")).toMatchSnapshot();
    });

    it("generate nice error message for invalid lambda expressions", () => {
      expect(compileError("(lambda)")).toMatchSnapshot();
      expect(compileError("(lambda 7)")).toMatchSnapshot();
      expect(compileError("(lambda 7 5)")).toMatchSnapshot();
      expect(compileError("(lambda (7) x)")).toMatchSnapshot();
      expect(compileError("(lambda (x 7) x)")).toMatchSnapshot();
      expect(compileError("(lambda (x x) x)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid let expressions", () => {
      expect(compileError("(let)")).toMatchSnapshot();
      expect(compileError("(let x 5)")).toMatchSnapshot();
      expect(compileError("(let (x 5) x)")).toMatchSnapshot();
    });

    it.skip("generate nice error message a let expression with odd numbrer of elements in the bindings", () => {
      // Those two tests cases have been commented out because the
      // error is coming from the reader, not the converter, so they
      // are not handled very nicely now.
      expect(compileError("(let {x} x)")).toMatchSnapshot();
      expect(compileError("(let {a b c} x)")).toMatchSnapshot();
    });

    it.skip("generate nice error message for invalid let expressions with duplicated variable names", () => {
      expect(compileError("(let {x 10 x 20} x)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid definitions", () => {
      expect(compileError("(define)")).toMatchSnapshot();
      expect(compileError("(define 5)")).toMatchSnapshot();
      expect(compileError("(define x)")).toMatchSnapshot();
      expect(compileError("(define 4 2)")).toMatchSnapshot();
      expect(compileError("(define x 10 23)")).toMatchSnapshot();
    });

    it("generate nice error message for conditionals", () => {
      expect(compileError("(if)")).toMatchSnapshot();
      expect(compileError("(if 5)")).toMatchSnapshot();
      expect(compileError("(if x 2)")).toMatchSnapshot();
      expect(compileError("(if x 2 3 0)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid records", () => {
      expect(compileError("{:x 10 | y :z 20}")).toMatchSnapshot();
    });

    it("generate nice error message for records with duplicated labels", () => {
      expect(compileError("{:x 10 :x 20}")).toMatchSnapshot();
    });

    it("generate nice error message for invalid type annotations", () => {
      expect(compileError("(the)")).toMatchSnapshot();
      expect(compileError("(the 3)")).toMatchSnapshot();
      expect(compileError("(the ID)")).toMatchSnapshot();
      expect(compileError("(the 3 3)")).toMatchSnapshot();
      expect(compileError("(the ID 3 3)")).toMatchSnapshot();
    });

    it("generate nice error message for bad type declarations", () => {
      expect(compileError("(type)")).toMatchSnapshot();
      expect(compileError("(type a)")).toMatchSnapshot();
      expect(compileError("(type A)")).toMatchSnapshot();
      expect(compileError("(type A 3)")).toMatchSnapshot();
      expect(compileError("(type a {})")).toMatchSnapshot();
      expect(compileError("(type ID {} 10)")).toMatchSnapshot();
      expect(compileError("(type 3 5)")).toMatchSnapshot();
      expect(compileError("(type ID (-> a a))")).toMatchSnapshot();
    });

    it("generate nice error messages for invalid exports", () => {
      expect(compileError("(export)")).toMatchSnapshot();
      expect(compileError("(export 1)")).toMatchSnapshot();
      expect(compileError("(export (+ 1 2))")).toMatchSnapshot();
      expect(compileError("(export 1 2 3)")).toMatchSnapshot();
    });

    it("generate nice errors for do-blocks", () => {
      expect(compileError("(do)")).toMatchSnapshot();
    });

    it("generaten nice errors for multiple-value-bind form", () => {
      expect(compileError("(multiple-value-bind)")).toMatchSnapshot();
      expect(compileError("(multiple-value-bind x)")).toMatchSnapshot();
      expect(compileError("(multiple-value-bind x y)")).toMatchSnapshot();
      expect(compileError("(multiple-value-bind (x) y)")).toMatchSnapshot();
    });
  });
});
