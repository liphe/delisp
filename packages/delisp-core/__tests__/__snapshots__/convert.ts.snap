// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Convert Error messages generate a nice error for some basic invalid syntax 1`] = `
"file:1:1: Empty list is not a function call
()
^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 1`] = `
"file:1:5: 'type' needs exactly 2 arguments, got 0
(type)
----^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 2`] = `
"file:1:7: 'type' needs exactly 2 arguments, got 1
(type a)
------^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 3`] = `
"file:1:7: 'type' needs exactly 2 arguments, got 1
(type A)
------^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 4`] = `
"file:1:9: Not a valid type
(type A 3)
--------^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 5`] = `
"file:1:7: 'a' is not a valid name for a type. Type names should start with a capital letter. Try 'A'?
(type a {})
------^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 6`] = `
"file:1:14: 'type' needs exactly 2 arguments, got 3
(type ID {} 10)
-------------^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 7`] = `
"file:1:7: 'type' expected a symbol as a name
(type 3 5)
------^"
`;

exports[`Convert Error messages generate nice error message for bad type declarations 8`] = `
"file:1:10: Type variable out of scope
(type ID (-> a a))
---------^"
`;

exports[`Convert Error messages generate nice error message for conditionals 1`] = `
"file:1:3: 'if' needs exactly 3 arguments, got 0
(if)
--^"
`;

exports[`Convert Error messages generate nice error message for conditionals 2`] = `
"file:1:5: 'if' needs exactly 3 arguments, got 1
(if 5)
----^"
`;

exports[`Convert Error messages generate nice error message for conditionals 3`] = `
"file:1:7: 'if' needs exactly 3 arguments, got 2
(if x 2)
------^"
`;

exports[`Convert Error messages generate nice error message for conditionals 4`] = `
"file:1:11: 'if' needs exactly 3 arguments, got 4
(if x 2 3 0)
----------^"
`;

exports[`Convert Error messages generate nice error message for invalid definitions 1`] = `
"file:1:7: 'define' needs exactly 2 arguments, got 0
(define)
------^"
`;

exports[`Convert Error messages generate nice error message for invalid definitions 2`] = `
"file:1:9: 'define' needs exactly 2 arguments, got 1
(define 5)
--------^"
`;

exports[`Convert Error messages generate nice error message for invalid definitions 3`] = `
"file:1:9: 'define' needs exactly 2 arguments, got 1
(define x)
--------^"
`;

exports[`Convert Error messages generate nice error message for invalid definitions 4`] = `
"file:1:9: 'define' expected a symbol
(define 4 2)
--------^"
`;

exports[`Convert Error messages generate nice error message for invalid definitions 5`] = `
"file:1:15: 'define' needs exactly 2 arguments, got 3
(define x 10 23)
--------------^"
`;

exports[`Convert Error messages generate nice error message for invalid lambda expressions 1`] = `
"file:1:7: 'lambda' is missing the argument list
(lambda)
------^"
`;

exports[`Convert Error messages generate nice error message for invalid lambda expressions 2`] = `
"file:1:9: Expected a list of arguments
(lambda 7)
--------^"
`;

exports[`Convert Error messages generate nice error message for invalid lambda expressions 3`] = `
"file:1:9: Expected a list of arguments
(lambda 7 5)
--------^"
`;

exports[`Convert Error messages generate nice error message for invalid lambda expressions 4`] = `
"file:1:10: A list of arguments should be made of symbols
(lambda (7) x)
---------^"
`;

exports[`Convert Error messages generate nice error message for invalid lambda expressions 5`] = `
"file:1:12: A list of arguments should be made of symbols
(lambda (x 7) x)
-----------^"
`;

exports[`Convert Error messages generate nice error message for invalid lambda expressions 6`] = `
"file:1:12: There is another argument with the same name
(lambda (x x) x)
-----------^"
`;

exports[`Convert Error messages generate nice error message for invalid let expressions 1`] = `
"file:1:4: 'let' is missing the bindings
(let)
---^"
`;

exports[`Convert Error messages generate nice error message for invalid let expressions 2`] = `
"file:1:6: 'let' bindings should be a map
(let x 5)
-----^"
`;

exports[`Convert Error messages generate nice error message for invalid let expressions 3`] = `
"file:1:6: 'let' bindings should be a map
(let (x 5) x)
-----^"
`;

exports[`Convert Error messages generate nice error message for invalid records 1`] = `
"file:1:8: '|' is not a valid field name. Fields should start with ':' (colon).
{:x 10 | y :z 20}
-------^"
`;

exports[`Convert Error messages generate nice error message for invalid type annotations 1`] = `
"file:1:4: 'the' is missing the type and value
(the)
---^"
`;

exports[`Convert Error messages generate nice error message for invalid type annotations 2`] = `
"file:1:6: 'the' is missing the expression
(the 3)
-----^"
`;

exports[`Convert Error messages generate nice error message for invalid type annotations 3`] = `
"file:1:7: 'the' is missing the expression
(the ID)
------^"
`;

exports[`Convert Error messages generate nice error message for invalid type annotations 4`] = `
"file:1:6: Not a valid type
(the 3 3)
-----^"
`;

exports[`Convert Error messages generate nice error message for invalid type annotations 5`] = `
"file:1:11: Too many arguments. 'the' should take two arguments.
(the ID 3 3)
----------^"
`;

exports[`Convert Error messages generate nice error message for records with duplicated labels 1`] = `
"file:1:8: Duplicated label
{:x 10 :x 20}
-------^"
`;

exports[`Convert Error messages generate nice error messages for invalid exports 1`] = `
"file:1:7: 'export' needs exactly 1 arguments, got 0
(export)
------^"
`;

exports[`Convert Error messages generate nice error messages for invalid exports 2`] = `
"file:1:9: 'export' expected a symbol
(export 1)
--------^"
`;

exports[`Convert Error messages generate nice error messages for invalid exports 3`] = `
"file:1:9: 'export' expected a symbol
(export (+ 1 2))
--------^"
`;

exports[`Convert Error messages generate nice error messages for invalid exports 4`] = `
"file:1:13: 'export' needs exactly 1 arguments, got 3
(export 1 2 3)
------------^"
`;

exports[`Convert Error messages generate nice errors for do-blocks 1`] = `
"file:1:3: empty body
(do)
--^"
`;

exports[`Convert Error messages generaten nice errors for multiple-value-bind form 1`] = `
"file:1:1: Ill-formed form. 

(multiple-value-bind (variable1 variable2 ...) 
    form 
  ...body)
(multiple-value-bind)
^"
`;

exports[`Convert Error messages generaten nice errors for multiple-value-bind form 2`] = `
"file:1:1: Ill-formed form. 

(multiple-value-bind (variable1 variable2 ...) 
    form 
  ...body)
(multiple-value-bind x)
^"
`;

exports[`Convert Error messages generaten nice errors for multiple-value-bind form 3`] = `
"file:1:1: Ill-formed form. 

(multiple-value-bind (variable1 variable2 ...) 
    form 
  ...body)
(multiple-value-bind x y)
^"
`;

exports[`Convert Error messages generaten nice errors for multiple-value-bind form 4`] = `
"file:1:1: Ill-formed form. 

(multiple-value-bind (variable1 variable2 ...) 
    form 
  ...body)
(multiple-value-bind (x) y)
^"
`;
